HWDataset.py

```diff
@@ -0,0 +1,90 @@
+from torch.utils.data import Dataset
+from torch.utils.data import DataLoader
+from pathlib import Path
+import math
+import numpy as np
+import plot_stroke
+import torch
+
+def collate_fn(samples):
+    batch_size=len(samples)
+    strokes = [s for s, text in samples]
+    texts = [text for s, text in samples]
+
+    max_stroke_len = max([len(s) for s in strokes])
+    num_stroke_elements=strokes[0].shape[1]
+    all_strokes = torch.zeros((max_stroke_len, batch_size, num_stroke_elements))
+    strokes_mask = torch.ones((max_stroke_len, batch_size), dtype=torch.bool)
+    for i in range(batch_size):
+        this_stroke_len = strokes[i].shape[0]
+        # pad
+        all_strokes[:this_stroke_len,i] = torch.tensor(strokes[i])
+        # mask
+        strokes_mask[this_stroke_len:,i] = 0
+
+    max_texts_length = max([len(t) for t in texts])
+    all_texts = torch.zeros((max_texts_length, batch_size), dtype=torch.int8)
+    texts_mask = torch.ones((max_texts_length, batch_size), dtype=torch.bool)
+    for i in range(batch_size):
+        this_texts_len = texts[i].shape[0]
+        # pad
+        all_texts[:this_texts_len,i] = torch.tensor(texts[i])
+        # mask
+        texts_mask[this_texts_len:,i] = 0
+
+    return all_texts, texts_mask, all_strokes, strokes_mask
+    
+class HandwritingDataset(Dataset):
+    def __init__(self, is_validation=False, validation_percentage=.1, data_dir=Path('./descript-research-test/data')):
+        super().__init__()
+        self.strokes = np.load(data_dir / 'strokes-py3.npy', allow_pickle=True)
+        sents = (data_dir / 'sentences.txt').read_text().splitlines()
+        self.sentences = [self.text2code(s) for s in sents]
+        assert len(self.sentences) == len(self.strokes)
+
+        num_validation = math.floor(len(self.strokes) * validation_percentage)
+        if is_validation:
+            self.sentences = self.sentences[-num_validation:]
+            self.strokes = self.strokes[-num_validation:]
+        else:
+            self.sentences = self.sentences[:-num_validation]
+            self.strokes = self.strokes[:-num_validation]
+
+    def char2index(self, ch):
+        return ord(ch)
```

Comment:
Hmm, not sure that I understand this line as I've never used `ord`. What do you think that we're trying to do with this?

Reply:
Convert chars to ints and back.  Ideally, we'd only use as many ints as there are different chars. ord/chr just go from ascii characters to their integer equivalent.

Reply:
I don't think that this works for what we want to do because we will be building an embedding table. Since we don't want to have unused rows, we want to have as many ints as there are different chars.

In practice, we usually build the set of characters with something like:
```[<S>, <E>, ] + sorted(list(set(all_text)))```

Reply:
This just converted chars to their ascii equivalent.  I've converted to using a smaller list of chars.